src/Microsoft.AspNetCore.Authentication/AuthenticationSchemeOptions.cs

```diff
@@ -36,5 +37,21 @@ public virtual void Validate(string scheme)
         /// If set, will be used as the service type to get the Events instance instead of the property.
         /// </summary>
         public Type EventsType { get; set; }
+
+        // Scheme forwarding properties
+
+        public string ForwardDefault { get; set; }
+
+        public string ForwardAuthenticate { get; set; }
+        public string ForwardChallenge { get; set; }
+        public string ForwardForbid { get; set; }
+        public string ForwardSignIn { get; set; }
+        public string ForwardSignOut { get; set; }
+
+        /// <summary>
+        /// Used to select a default scheme to target based on the request.
+        /// </summary>
+        public Func<HttpContext, string> ForwardDefaultSelector { get; set; }
```

Comment:
The selector also only makes sense for the virtual handler. Using it on any other handler prevents that handler from ever working.

Reply:
Every handler is a virtual handler now, so a better way to think about this is to think of the forwarding as a selective opt out of that handlers functionality, so lets say you have a google handler that you want to do something else for /custom/ requests, that's what you would do.

That's basically what this feature is, if you set the default selector and it returns something non null, that's basically saying to the handler to get out of the way and let another scheme handle everything.

Reply:
A more concrete example would be for apps with a single social auth, lets say google.

That could be setup now without a 3rd virtual scheme as:
```
   services.AddAuthentication("Cookies").AddCookie(o => o.ForwardChallenge = "Google").AddGoogle()
```

Reply:
Note: that's basically what the new `services.AddAuthentication().AddGoogleSignIn()` sugar will setup by default (add a cookie if it doesn't exist, and then point it at "Google" for challenges, the equivalent for AddIdentity())

Reply:
So an app that wanted to do API vs Cookies would make Cookies the default and set the ForwardDefaultSelector on CookieAuthOptions?

How about ForwardDefault? Is that ever useful outside of the virtual handler?

Reply:
I think its still useful as the idea is all handlers are now able to be used as virtual schemes, but I see the problem, now that this is merged, we need a way to opt out of the ForwardDefault.

So I think the tweak is simply to set ForwardAuthenticate to yourself

```
   services.AddAuthentication("Google").AddGoogle(o => {
      o.ForwardDefault = "Cookies"
      o.ForwardChallenge = "Google"
   }
```

This would block falling back to ForwardDefault, and I can add a scheme name check to call into HandleXyzAsync() if forward is targeting self.